### PR TITLE
Optimize redundant adds into fixed size priority queue for topN computation.

### PR DESCRIPTION
**Pull request by mayankshriv**
The current code to compute topN adds elements into priority queue
unconditionally, only to de-queue immediately if queue size exceeds N.
One optimization here is to compare the new element against top of the
queue, and add only if it is better than the top value. Doing so
provides about 10% latency improvement for cases with large number of
groups (100k groups or more, before pruning).

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java [hunk 1]
```diff
@@ -115,9 +115,15 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
           if (priorityQueue == null) {
             trimmedResults.get(i).put(groupKey, intermediateResults[i]);
           } else {
-            priorityQueue.add(new GroupKeyResultPair(groupKey, (Comparable) intermediateResults[i]));
-            if (priorityQueue.size() > _trimSize) {
-              priorityQueue.poll();
+            GroupKeyResultPair newValue = new GroupKeyResultPair(groupKey, (Comparable) intermediateResults[i]);
+            if (priorityQueue.size() >= _trimSize) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: "=="?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java [hunk 2]
```diff
@@ -171,9 +177,16 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
       for (Map.Entry<String, Comparable> entry : finalResultMap.entrySet()) {
         String groupKey = entry.getKey();
         Comparable finalResult = entry.getValue();
-        priorityQueue.add(new GroupKeyResultPair(groupKey, finalResult));
-        if (priorityQueue.size() > _groupByTopN) {
-          priorityQueue.poll();
+
+        if (priorityQueue.size() >= _groupByTopN) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: Same here

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java [hunk 3]
```diff
@@ -171,9 +177,16 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
...
+          GroupKeyResultPair topValue = priorityQueue.peek();
+          GroupKeyResultPair newValue = new GroupKeyResultPair(groupKey, finalResult);
+          if (priorityQueue.comparator().compare(topValue, newValue) < 0) {
+            priorityQueue.poll();
+            priorityQueue.add(new GroupKeyResultPair(groupKey, finalResult));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: newValue

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java [hunk 4]
```diff
@@ -171,9 +177,16 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
...
+          }
+        } else {
+          priorityQueue.add(new GroupKeyResultPair(groupKey, finalResult));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: newValue

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: Can _trimSize be 0? If yes, then we need to first add then pool

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: _trimSize is set to a min of 5000 in the current code. Nonetheless, I have added a null check against topValue (will happen when the queue is empty).

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: I feel >= is more resilient to future changes.

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java (hunk 2 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: Yeah, same response :-)

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java (hunk 3 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: Addressed

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java (hunk 4 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: Addressed

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java [hunk 5]
```diff
@@ -171,9 +177,16 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
       for (Map.Entry<String, Comparable> entry : finalResultMap.entrySet()) {
         String groupKey = entry.getKey();
         Comparable finalResult = entry.getValue();
-        priorityQueue.add(new GroupKeyResultPair(groupKey, finalResult));
-        if (priorityQueue.size() > _groupByTopN) {
-          priorityQueue.poll();
+
+        GroupKeyResultPair newValue = new GroupKeyResultPair(groupKey, finalResult);
+        if (priorityQueue.size() >= _groupByTopN) {
```
Comment (snleee) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java]: This code looks the same as the code above. We can probably take out these as a separate function to reuse the code. i.e. `addToPriorityQueue(priorityQueue, maxSize)`

**Pull request merged**